fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinReorderHelper.java

```diff
@@ -0,0 +1,168 @@
+                    top.put(entry.getKey(), entry.getValue());
+                }
+            }
+        }
+        return ImmutableList.of(bottom, top);
+    }
+
+    /*
+     * split join output cols into three parts:
+     * 1. cols only ref left child cols
+     * 2. cols only ref right child cols
+     * 3. cols ref both left and right children cols
+     * Actually there may be some constants cols, we don't need to consider them here.
+     */
+    public static List<ColumnRefSet> splitJoinOutputCols(OptExpression joinExpr) {
+        LogicalJoinOperator joinOperator = (LogicalJoinOperator) joinExpr.getOp();
+        OptExpression left = joinExpr.inputAt(0);
+        OptExpression right = joinExpr.inputAt(1);
+        ColumnRefSet leftCols = new ColumnRefSet();
+        ColumnRefSet rightCols = new ColumnRefSet();
+        ColumnRefSet refBothChildCols = new ColumnRefSet();
+
+        Projection projection = joinOperator.getProjection();
+        if (projection == null) {
+            leftCols.union(left.getOutputColumns());
+            rightCols.union(right.getOutputColumns());
+        } else {
+            for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : projection.getColumnRefMap().entrySet()) {
+                ColumnRefOperator columnRefOperator = entry.getKey();
+                ScalarOperator scalarOperator = entry.getValue();
+                if (scalarOperator.getUsedColumns().isEmpty()) {
+                    // do nothing
+                } else if (left.getOutputColumns().containsAll(scalarOperator.getUsedColumns())) {
+                    leftCols.union(columnRefOperator.getId());
+                } else if (right.getOutputColumns().containsAll(scalarOperator.getUsedColumns())) {
+                    rightCols.union(columnRefOperator.getId());
+                } else {
+                    refBothChildCols.union(columnRefOperator.getId());
+                }
+            }
+        }
+        return Lists.newArrayList(leftCols, rightCols, refBothChildCols);
+    }
+
+
+    public static void splitCondition(
+            ScalarOperator onCondition, ColumnRefSet columnRefSet,
+            List<ScalarOperator> containsAll, List<ScalarOperator> nonContainsAll) {
+        List<ScalarOperator> conjuncts = Utils.extractConjuncts(onCondition);
+        for (ScalarOperator conjunct : conjuncts) {
+            if (columnRefSet.containsAll(conjunct.getUsedColumns())) {
+                containsAll.add(conjunct);
+            } else {
+                nonContainsAll.add(conjunct);
+            }
+
+        }
+    }
+
+    public static Map<ColumnRefOperator, ScalarOperator> mergeProject(Map<ColumnRefOperator, ScalarOperator> project,
+                                      Map<ColumnRefOperator, ScalarOperator> inputProject) {
+        Map<ColumnRefOperator, ScalarOperator> newProject = Maps.newHashMap(inputProject);
+
+        ReplaceColumnRefRewriter rewriter = new ReplaceColumnRefRewriter(inputProject);
```

Comment:
Why rewrite it before merge ?

Reply:
For example, the splitted projection is 4:col -> 3:col and the child has a projection of 3:col -> abs(v1), we need rewrite the split projection to 4:col -> abs(v1) then merge it into the child's projection